contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,557 @@
+// returns the new list
+message ListWatchResponse {
+	repeated Device devices = 1;
+}
+
+// The list of be added to the CRI spec
+message DeviceRuntimeSpec {
+	// List of environment variable to set in the container.
+	repeated KeyValue envs = 1;
+	// Mounts for the container.
+	repeated Mount mounts = 2;
+}
+
+// E.g:
+// struct Device {
+//    ID: "GPU-fef8089b-4820-abfc-e83e-94318197576e",
+//    State: "Healthy",
+//}
+message Device {
+	string id = 1;
+	string state = 2;
+
+	DeviceRuntimeSpec spec = 3;
+}
+```
+
+We are expecting the State property to have some well known values:
+```
+const (
+	// Healthy means that the device is healty
+	Healty = "Healthy"
+	// UnHealthy means that the device is unhealty
+	Unhealthy = "Unhealty"
+)
+```
+
+### HealthCheck and Failure Recovery
+
+We want Kubelet as well as the Device Plugins to recover from failures
+that may happen on any side of this protocol.
+
+At the communication level, gRPC is a very strong piece of software and
+is able to ensure that if failure happens it will try it's best to recover
+through exponential backoff reconnection and Keep alive checks.
+
+The proposed mechanism intends to replace any device specific handling in
+Kubelet. Therefore in general, device plugin failure or upgrade means that
+Kubelet is not able to accept any pod requesting a Device until the upgrade
+or failure finishes.
+
+If a device fails, the Device Plugin should signal that through the
+`ListAndWatch` gRPC stream. We then expect Kubelet to fail the Pod.
+
+If any Device Plugin fails the behavior we expect depends on the task Kubelet
+is performing:
+* In general we expect Kubelet to remove any devices that are owned by the failed
+  device plugin from the resources advertised by the Node status.
+* We however do not expect Kubelet to fail or restart any pods or containers
+  running that are using these devices.
+* If Kubelet is in the process of allocating a device, then it should fail
+  the container process.
+
+If the Kubelet fails or restarts, we expect the Device Plugins to know about
+it through gRPC's Keep alive feature and try to reconnect to Kubelet.
+
+When Kubelet fails or restarts it should know what are the devices that are
+owned by the different containers and be able to rebuild a list of available
+devices.
+We are expecting to implement this through a checkpointing mechanism that Kubelet
+would write and read from.
+
+
+## API Changes
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server as part of the kubelet node update current protocol.
+
+We will be using extended resources to schedule, trigger and advertise these
+Devices.
+When a Device plugin registers two `foo-device` the node status will be
+updated to advertise 2 `extensions.kubernetes.io/foo-device`.
+
+If a user wants to trigger the device plugin he only needs to request this
+through the same mechanism as OIRs in his Pod Spec.
+
+*Future*:
+
```

Comment:
If we list it here, people would get the impression that this is the decided way to go. But that is not the reality. We are still debating this even inside the resource management team. You can put it in a separate PR proposal for general discussion, but I would like this PR to cover the design and the plans we have agreed on, so that we can merge it soon.